src/server/read_req.rs

```diff
@@ -187,15 +227,31 @@ where
                 if let Ok(Packet::Ack(recved_block_id)) =
                     Packet::decode(&buf[..len])
                 {
-                    if recved_block_id == block_id {
-                        return Ok(());
+                    let window_end = window_base.wrapping_add(window_len);
+
+                    if window_end > window_base {
+                        // window_end did not wrap
+                        if recved_block_id >= window_base && recved_block_id < window_end {
+                            // number of blocks acked
+                            return Ok(recved_block_id-window_base+1u16);
+                        }
+                        else {
+                            trace!("Unexpected ack packet {recved_block_id}, window_base: {window_base}, window_len: {window_len}");
+                        }
+                    }else {
+                        // window_end wrapped
+                        if recved_block_id >= window_base {
+                            return Ok(1u16 + (recved_block_id - window_base));
+                        } else if recved_block_id < window_end {
+                            return Ok(1u16 + recved_block_id + window_len-window_end);
```

Comment:
You forgot the parenthesis here. It can overflow without them.

```suggestion
                            return Ok(1u16 + recved_block_id + (window_len - window_end));
```